primitives/src/merkle_tree/namespaced_merkle_tree.rs

```diff
@@ -181,10 +209,460 @@ where
+    ) -> Result<Self, PrimitivesError> {
+        let mut namespace_ranges: HashMap<N, Range<u64>> = HashMap::new();
+        let mut max_namespace = <N as Namespace>::min();
+        let mut leaves = Vec::new();
+        for (idx, elem) in elems.into_iter().enumerate() {
+            let ns = elem.borrow().get_namespace();
+            let idx: u64 = idx.try_into().unwrap();
+            if ns < max_namespace {
+                return Err(PrimitivesError::InconsistentStructureError(
+                    "Namespace leaves must be pushed in sorted order".into(),
+                ));
+            }
+            match namespace_ranges.entry(ns) {
+                Entry::Occupied(entry) => {
+                    entry.into_mut().end = idx + 1;
+                },
+                Entry::Vacant(entry) => {
+                    entry.insert(idx..idx + 1);
+                },
+            }
+            max_namespace = ns;
+            leaves.push(elem);
+        }
+        let inner = <InnerTree<E, H, T, N, Arity> as MerkleTreeScheme>::from_elems(height, leaves)?;
+        Ok(NMT {
+            inner,
+            namespace_ranges,
+        })
+    }
+
+    fn height(&self) -> usize {
+        self.inner.height()
+    }
+
+    fn capacity(&self) -> num_bigint::BigUint {
+        self.inner.capacity()
+    }
+
+    fn num_leaves(&self) -> u64 {
+        self.inner.num_leaves()
+    }
+
+    fn commitment(&self) -> Self::Commitment {
+        self.inner.commitment()
+    }
+
+    fn lookup(
+        &self,
+        pos: impl Borrow<Self::Index>,
+    ) -> super::LookupResult<Self::Element, Self::MembershipProof, ()> {
+        self.inner.lookup(pos)
+    }
+
+    fn verify(
+        root: impl Borrow<Self::NodeValue>,
+        pos: impl Borrow<Self::Index>,
+        proof: impl Borrow<Self::MembershipProof>,
+    ) -> Result<VerificationResult, PrimitivesError> {
+        <InnerTree<E, H, T, N, Arity> as MerkleTreeScheme>::verify(root, pos, proof)
+    }
+}
+
+impl<E, H, Arity, N, T> AppendableMerkleTreeScheme for NMT<E, H, Arity, N, T>
+where
+    H: DigestAlgorithm<E, u64, T> + BindNamespace<E, u64, T, N>,
+    E: Element + Namespaced<Namespace = N>,
+    T: NodeValue,
+    N: Namespace,
+    Arity: Unsigned,
+{
+    fn extend(
+        &mut self,
+        elems: impl IntoIterator<Item = impl core::borrow::Borrow<Self::Element>>,
+    ) -> Result<(), PrimitivesError> {
+        // TODO(#314): update namespace metadata
+        self.inner.extend(elems)
+    }
+
+    fn push(
+        &mut self,
+        elem: impl core::borrow::Borrow<Self::Element>,
+    ) -> Result<(), PrimitivesError> {
+        // TODO(#314): update namespace metadata
+        self.inner.push(elem)
+    }
+}
+
+/// Indicates whether the namespace proof represents a populated set or an empty
+/// set
+#[derive(Clone, Debug)]
+enum NamespaceProofType {
+    Presence,
+    Absence,
+}
+
+#[derive(Clone, Debug)]
+/// Namespace Proof
+pub struct NaiveNamespaceProof<E, T, Arity, N, H>
```

Comment:
Should we also derive `CanonicalSerialize` and `CanonicalDeserialize` to this type?
We might also want `Eq`, `PartialEq`, etc.

Reply:
yeah - I guess we can derive Serialize/Deserialize and have canonical serialize bounds for `E, T`?

Reply:
Not sure about `Canonical*` but we will definitely need serde `Serialize` and `Deserialize`, for the query service

Reply:
done 👍